src/io/ctrees_utils.c

```diff
@@ -112,27 +114,38 @@ int64_t read_locations(const char *filename, const int64_t ntrees, struct locati
             XASSERT(ntrees_found < ntrees, EXIT_FAILURE,
                     "ntrees=%"PRId64" should be less than ntrees_found=%"PRId64"\n",
                     ntrees, ntrees_found);
-            int nitems = sscanf(buffer, "%"SCNd64" %d %"SCNd64 "%s",
+            const int nitems = sscanf(buffer, "%"SCNd64" %d %"SCNd64 "%s",
                                 &locations->treeid, &locations->fileid, &locations->offset, linebuf);
+            XASSERT(nitems == nitems_expected, EXIT_FAILURE, "Expected to parse %d items but the scanf produced %d items instead.\n"
+                    "The bufound `%s' in the buffer\n", nitems_expected, nitems, buffer);
 
             XASSERT(locations->offset >= 0, INVALID_VALUE_READ_FROM_FILE,
                     "offset=%"PRId64" for ntree =%"PRId64" must be positive.\nFile = `%s'\nbuffer = `%s'\n",
                     locations->offset, ntrees_found, filename, buffer);
 
+            XASSERT(locations->fileid >= 0, INVALID_VALUE_READ_FROM_FILE,
+                    "locations->fileid=%d for ntree =%"PRId64" must be positive.\nFile = `%s'\nbuffer = `%s'\n",
+                    locations->fileid, ntrees_found, filename, buffer);
             const size_t fileid = locations->fileid;
             if(fileid > files_fd->nallocated) {
                 numfiles_allocated *= 2;
                 int *new_fd = realloc(files_fd->fd, numfiles_allocated*sizeof(files_fd->fd[0]));
                 XASSERT(new_fd != NULL, MALLOC_FAILURE, "Error: Could not re-allocate memory of %zu bytes to hold %"PRIu32" file descriptors\n",
                         numfiles_allocated*sizeof(files_fd->fd[0]), numfiles_allocated);
 
+                void *new_numtrees = realloc(files_fd->numtrees_per_file, numfiles_allocated*sizeof(files_fd->numtrees_per_file[0]));
+                XASSERT(new_numtrees != NULL, MALLOC_FAILURE, "Error: Could not re-allocate memory of %zu bytes to hold %"PRIu32" items containing 64-bit integers\n",
+                        numfiles_allocated*sizeof(files_fd->numtrees_per_file[0]), numfiles_allocated);
+                
                 files_fd->fd = new_fd;
+                files_fd->numtrees_per_file = new_numtrees;
                 for(uint32_t i=files_fd->nallocated;i<numfiles_allocated;i++) {
                     files_fd->fd[i] = -1;
+                    files_fd->numtrees_per_file[i] = 0;
                 };
                 files_fd->nallocated = numfiles_allocated;
             }
-            XASSERT(fileid < files_fd->nallocated, EXIT_FAILURE, "Error: Trying to open fileid = %zu but only %d"
+            XRETURN(fileid < files_fd->nallocated, EXIT_FAILURE, "Error: Trying to open fileid = %zu but only %u"
```

Comment:
(This is me deleting a comment because I accidentally told Firefox to not allow additional dialogs which stops me from deleting comments...)